modules/ppcp-agentic-commerce/src/CartValidation/CouponValidator/CouponValidator.php

```diff
@@ -0,0 +1,536 @@
+ */
+
+declare(strict_types=1);
+
+namespace WooCommerce\PayPalCommerce\AgenticCommerce\CartValidation\CouponValidator;
+
+use WC_Coupon;
+use WC_Discounts;
+use WP_Error;
+use WooCommerce\PayPalCommerce\AgenticCommerce\CartValidation\ValidatorInterface;
+use WooCommerce\PayPalCommerce\AgenticCommerce\Helper\CartHelper;
+use WooCommerce\PayPalCommerce\AgenticCommerce\Schema\Coupon;
+use WooCommerce\PayPalCommerce\AgenticCommerce\Schema\PayPalCart;
+use WooCommerce\PayPalCommerce\AgenticCommerce\Validation\CouponInvalid;
+
+/**
+ * Validates coupons for Agentic Commerce using WooCommerce's native validation.
+ */
+class CouponValidator implements ValidatorInterface {
+
+	/**
+	 * Context builder for building validation context data.
+	 *
+	 * @var ContextBuilder
+	 */
+	private ContextBuilder $context_builder;
+
+	/**
+	 * Discount calculator for coupon discount amounts.
+	 *
+	 * @var DiscountCalculator
+	 */
+	private DiscountCalculator $discount_calculator;
+
+	/**
+	 * Resolution builder for building resolution options.
+	 *
+	 * @var ResolutionBuilder
+	 */
+	private ResolutionBuilder $resolution_builder;
+
+	/**
+	 * Error message patterns for mapping WC errors to PayPal issue types.
+	 *
+	 * WooCommerce's WC_Discounts always returns WP_Error with code 'invalid_coupon'.
+	 * The actual error type is determined by the error message content.
+	 * Patterns are checked in order - more specific patterns should come first.
+	 */
+	private const MESSAGE_PATTERNS = array(
+		// Existence checks.
+		'does not exist'                              => 'COUPON_NOT_EXIST',
+		'cannot be applied because it does not exist' => 'COUPON_NOT_EXIST',
+
+		// Expiration.
+		'has expired'                                 => 'COUPON_EXPIRED',
+
+		// Usage limits.
+		'usage limit'                                 => 'USAGE_LIMIT_EXCEEDED',
+		'reached its usage limit'                     => 'USAGE_LIMIT_EXCEEDED',
+
+		// Spend requirements.
+		'minimum spend'                               => 'MINIMUM_ORDER_NOT_MET',
+		'maximum spend'                               => 'MAXIMUM_ORDER_EXCEEDED',
+
+		// Product/category restrictions - check specific patterns first.
+		'not applicable to the products:'             => 'COUPON_NOT_APPLICABLE',
+		'not applicable to the categories:'           => 'COUPON_NOT_APPLICABLE',
+		'not applicable to selected products'         => 'COUPON_NOT_APPLICABLE',
+		'not valid for sale items'                    => 'COUPON_NOT_APPLICABLE',
+		'not applicable'                              => 'COUPON_NOT_APPLICABLE',
+
+		// Email restrictions.
+		'not yours'                                   => 'COUPON_EMAIL_RESTRICTED',
+		'does not belong'                             => 'COUPON_EMAIL_RESTRICTED',
+
+		// Already applied.
+		'already applied'                             => 'COUPON_ALREADY_APPLIED',
+		'already been applied'                        => 'COUPON_ALREADY_APPLIED',
+	);
```

Comment:
Blocker: This likely works in the initial version (US shops with US customers), but error messages are localized by WooCommerce, making this pattern matching only reliable for english shops

https://woocommerce.github.io/code-reference/files/woocommerce-includes-class-wc-coupon.html#source-view.1060 →  `esc_html__(...)`